fix(popover): DP-162957 reinitialize the tippy when the anchor element changes

noellelevydialpad opened the pull request:
# PR Title
reinitialize the tippy when the anchor element changes

## Obligatory GIF (super important!)

![Obligatory GIF](https://media3.giphy.com/media/v1.Y2lkPTc5MGI3NjExYmh3cHAxNnA4bWE2ZnZtbmduY3czdXdrOWRkNmFhNDZsMDlra2dtZSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/fkLMC52ghhIUU/giphy.gif)

## :hammer_and_wrench: Type Of Change

These types will increment the version number on release:

- [x] Fix
- [ ] Feature
- [ ] Performance Improvement
- [ ] Refactor

These types will not increment the version number, but will still deploy to documentation site on release:

- [ ] Documentation
- [ ] Build system
- [ ] CI
- [ ] Style (code style changes, not css changes)
- [ ] Test
- [ ] Other (chore)

## :book: Jira Ticket

https://dialpad.atlassian.net/browse/DP-162957

## :book: Description

Use a mutationObserver to notice when the DOM element rendered into the content slot changes. This element is used to anchor the tippy, and if it has become detached from the DOM, tippy will throw an error when trying to find the element's parent.

As a bonus, this also makes sure to call destroy on this.tip before replacing it, this is possibly fixing a memory leak.

## :bulb: Context

In the native app, we wrap DtRecipeSettingsMenuButton in a DtPopover. When an app update becomes available, the root DOM element rendered by DtRecipeSettingsMenuButton changes, this breaks the tippy binding and clicking the menu just throws an error.

This does not seem to happen in vue 2, I suspect vue 2 is more sloppy about when it reuses DOM elments.

## :pencil: Checklist

<!--- Tick or place an `x` in all of the checkboxes that apply -->
<!--- Remove checkboxes that do not apply -->

For all PRs:

- [x] I have ensured no private Dialpad links or info are in the code or pull request description (Dialtone is a public repo!).
- [x] I have reviewed my changes.
- [x] I have added all relevant documentation.
- [x] I have considered the performance impact of my change.

For all Vue changes:

- [x] I have added / updated unit tests.
- [x] I have made my changes in Vue 2 and Vue 3. Note: you may sync your changes from Vue 2 to Vue 3 (or vice versa) using the `./scripts/dialtone-vue-sync.sh` script. Read docs here: [Dialtone Vue Sync Script](../packages/dialtone-vue3/.github/CONTRIBUTING.md#dialtone-vue-sync-script)
- [ ] I have validated components with a screen reader.
- [ ] I have validated components keyboard navigation.

## :crystal_ball: Next Steps

Similar changes should be made in other places we anchor a tippy to slot content, in particular in DtTooltip.

noellelevydialpad committed:
reinitialize the tippy when the anchor element changes

noellelevydialpad committed:
sync changes to vue2

iropolo commented:
I wouldn't touch the vue2 version if it is working as expected.
If we already verified it works the same Im okay with it.

iropolo reviewed (approved):
lgtm.

percy/dialtone failing is expected, approve the change before merge.

iropolo commented:
It is maybe related to https://dialpad.atlassian.net/browse/DP-164108

noellelevydialpad committed:
reproduce the bug in vue2 by forcing it to switch dom elements

this confirms my theory that vue2 was being sloppy about reusing dom
trees when different components resulted in similar trees.

noellelevydialpad commented:
> I wouldn't touch the vue2 version if it is working as expected. If we already verified it works the same Im okay with it.

I actually managed to reproduce the bug in vue2, it just doesn't happen to repro with DtRecipeSettingsMenuButton in vue2 because it sloppily reused the dom node between two different instances of DtButton. If I force it to swap out with a `key` attr, I get the same result as in vue3.

iropolo commented:
> > I wouldn't touch the vue2 version if it is working as expected. If we already verified it works the same Im okay with it.
> 
> I actually managed to reproduce the bug in vue2, it just doesn't happen to repro with DtRecipeSettingsMenuButton in vue2 because it sloppily reused the dom node between two different instances of DtButton. If I force it to swap out with a `key` attr, I get the same result as in vue3.

ohhh... make sense, yeah a clean up would be okay then.

braddialpad commented on packages/dialtone-vue2/components/popover/popover.test.js:
```diff
@@ -333,4 +334,55 @@ describe('DtPopover Tests', () => {
+      let popoverWindow = wrapper.findComponent({ ref: 'popover' }).findComponent({ ref: 'content' });
+      expect(popoverWindow.isVisible()).toBe(false);
+      wrapper.setProps({ open: true});
+      await nextTick();
```
nit: you can just await setProps instead of doing nextTick

braddialpad reviewed (approved):
Thanks for this!

noellelevydialpad committed:
await setProps
